### DISCUSSION TITLE
Lesson 7 testOnlyOwnerCanWithdraw()

### DISCUSSION THREAD

**Discussion by dimitar-dermendzhiev** [Cyfrin/foundry-full-course-cu#1777] [General]
Should the following test pass or fail? I don't use the ```vm.prank(USER)``` cheat code in this case and if I understand correct this test as it is should fail because the ```fundMe.withdraw()``` is being called by the DefaultSender who is also the owner of the fundMe contract. I get this test to pass as it is but I don't understand why it passes. 

```Solidity 
    function testOnlyOwnerCanWithdraw() public {
        vm.prank(USER);
        fundMe.fund{value: SEND_VALUE}();

        vm.expectRevert();
        fundMe.withdraw();
    }
  ```
  
  Another thing I want to ask is the order of ```vm.prank(USER)```  and ```vm.expectRevert()``` matters  since my understanding is that it doesn't matter as they skip one another.
  
  ```Solidity 
    function testOnlyOwnerCanWithdraw() public {
        vm.prank(USER);
        fundMe.fund{value: SEND_VALUE}();

        vm.expectRevert();
        vm.prank(USER);
        fundMe.withdraw();
    }
  ```

**Comment by urosognjenovic** [comment]
Your logic is on point. The test passes because you're expecting it to fail with the line ```vm.expectRevert();``` and it is called by the contract FundMeTest (or whichever contract the function testOnlyOwnerCanWithdraw is in).

The order of ```vm.prank(USER)``` and ```vm.expectRevert()``` doesn't matter because they are both cheatcodes.

**Reply by dimitar-dermendzhiev** [reply]
Shouldn't the test fail though? If I replace the ```vm.expectRevert()``` with ```vm.assertEq(fundMe.getOwner(), msg.sender)``` I have the following code:

```Solidity
function testOnlyOwnerCanWithdraw() public {
        vm.prank(USER);
        fundMe.fund{value: SEND_VALUE}();

        vm.assertEq(fundMe.getOwner(), msg.sender);
        fundMe.withdraw();
    }
```

The assertion seems to pass since the stack trace shows this:
```VM::assertEq(DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38], DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38]```

However for some reason ```fundMe.withdraw();``` still reverts ```← [Revert] FundMe__NotOwner()``` and I don't understand why is that since the DefaultSender is also the owner and should be able to process the withdraw.

I suspect that DefaultSender has some weird behavior which causes me headaches to understand.

**Reply by urosognjenovic** [reply]
```fundMe.withdraw()``` is called by the FundMeTest contract which is not the owner of the fundMe contract (I made a mistake above). That's why the test passes. ```msg.sender```, which is the owner of the fundMe contract, calls the TestFundMe contract which then calls the withdraw function.

**Reply by dimitar-dermendzhiev** [reply]
But why the stack trace shows that ```msg.sender``` and ```owner``` are both the same DefaultSender address ```0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38```

When I ```forge test --mt testOnlyOwnerCanWithdraw```
I understand the flow as:
```msg.sender``` (DefaultSender) creates  ```FundMeTest``` and becomes its owner
```FundMeTest``` becomes the owner of ```DeployFundMe``` from the ```SetUp()```
```msg.sender``` (DefaultSender) is owner of ```FundMe``` since he initialize the contract by calling ```deployFundMe.run()``` in the ```SetUp()``` 
```msg.sender``` calls ```fundMe.withdraw();``` in ```testOnlyOwnerCanWithdraw()```
```vm.expectRevert();``` does not revert and test should fail unless I use ```vm.prank(USER)```

**Reply by EngrPips** [reply]
Hello @dimitar-dermendzhiev, The test assertion passed, but the withdrawal failed because you did not prank the owner when you called the withdrawal function, as shown in your test below. Please note that the `defaultSender` in the Foundry test suite is not the address of your test contract but a default address that Foundry always uses whenever you don't explicitly provide a private key for a transaction. To clarify further, you can `console.log(fundMe.getOwner)` and `console.log(address.this)` inside your test and see that they are not the same.

```js
function testOnlyOwnerCanWithdraw() public {
        vm.prank(USER);
        fundMe.fund{value: SEND_VALUE}();

        vm.assertEq(fundMe.getOwner(), msg.sender);
        // There is no prank whatsoever here.
        fundMe.withdraw();
    }
```
